spark/src/main/java/org/apache/iceberg/spark/source/Reader.java

```diff
@@ -396,7 +397,15 @@ public void close() throws IOException {
       // schema or rows returned by readers
       Schema finalSchema = expectedSchema;
       PartitionSpec spec = task.spec();
-      Set<Integer> idColumns = spec.identitySourceIds();
+
+      Set<Integer> idColumns = Sets.newHashSet();
+      for (Integer i : spec.identitySourceIds()) {
```

Comment:
can you add a comment on why this check is needed?

Reply:
With this we check the nested partitions and we removed them from the right join side. We do this because the the left join already has the nested values inside because that spark doesn't remove remove nested fields even though you specify `partitionBy("nested.field")`.

So, in the case of doing `partitionBy("nested.field")` on a simple spark data frame and writing it as parquet, the result is thee fact that the file paths contains `.../nested.field=<value>/...` while the nested field is still present inside the data itself (inside the `.parquet` file).

Reply:
I think I see. Sounds like you're decoding the data written in data files, which keeps the existing structure. Is that right?

I'd still prefer to use the `ConstantReader` approach. Iceberg writes partition values to data files, but files imported from Hive tables -- which we do support -- won't have those values and those columns will be missing from the final data.

Reply:
@rdblue I'll try use a unit test over Hive to see the issue you are describing.

On the other hand I tried joining nested fields using the `JoinedRow` construct but it didn't suffice. 

Given the following original row:
```json
{
  "id": 1,
  "otherField": "otherVal", 
  "nestedData": { 
    "nestedKey1": "val1",  
    "nestedKey2": "val2" 
  } 
}
```
And writing it as Iceberg with the following 2 partitions: `identity(id)` and `identity(nestedData.nestedKey1)`.

Reading it back with Iceberg Reader would mean creating the row as a join between the partition values (ie: `1` and `val1`) and data (ie: `otherVal` and `val2`).

So, joining the following partition values
```json
{ 
  "id": 1, 
  "nestedData": {
    "nestedKey1": "val1" 
  } 
}
```
and data
```json
{ 
  "otherField": "otherVal", 
  "nestedData": { 
    "nestedKey2": "val2" 
  }
}
```
would end as 
```json
{ 
  "id": 1, 
  "nestedData": { 
    "nestedKey1": "val1" 
  },  
  "otherField": "otherVal", 
  "nestedData": { 
    "nestedKey2": "val2" 
  } 
}
```
The above is from what I've seen while putting together this fix.

But we would like it like this:
```json
{ 
  "id": 1, 
  "otherField": "otherVal", 
  "nestedData": { 
    "nestedKey1": "val1",  
    "nestedKey2": "val2" 
  } 
}
```
Right?

I don't know how the `ConstantReader` would help composing back that `nestedData` field as it originally was.

Reply:
> Iceberg writes partition values to data files, but files imported from Hive tables -- which we do support -- won't have those values and those columns will be missing from the final data

@rdblue can you point me to any unit test or snippet of code showing that flow -- loading Hive table data? I'm asking so I can have a starting point as my experience with Hive is not at the same level with Parquet. Thank you.

Reply:
Sorry for the late reply!

Partitioned tables converted using the [`importSparkTable`](https://github.com/apache/incubator-iceberg/blob/master/spark/src/main/scala/org/apache/iceberg/spark/SparkTableUtil.scala#L365-L368) won't have partition columns embedded in the file.

But that's not the only reason to use the `ConstantReader` approach. As I mentioned, using a `ConstantReader` will allow us to produce the schema that is requested by Spark without using a `JoinedRow` and a projection. The projection is expensive, so fixing this using a `ConstantReader` and removing the projection is a much better solution.